Improves the program derived addresses content (#380)
solana-foundation/developer-content: content/courses/native-onchain-development/program-derived-addresses.md

storage and retrieval.

PDAs serve two main functions:

1. Provide a deterministic way to find a given item of data for a program
2. Authorize the program that owns a PDA to sign on the PDAs behalf, just like a
   user signs for their own account using their secret key.

This lesson will focus on using PDAs to find and store data. We'll discuss
signing with a PDA more thoroughly in a future lesson, where we will cover
Cross-Program Invocations (CPIs).

### Finding PDAs

Technically, PDAs are _found_ or _derived_ based on a program ID and one or more
input seeds.

Unlike other Solana accounts, PDAs are not public keys and don't have secret
keys. Since public keys are on Solana's Ed25519 curve, PDAs are sometimes called
'off curve addresses'.

PDAs are found using a hashing function that deterministically generates a PDA
using the program ID and seeds. Both Solana frontend and backend code can
determine an address using the program ID and seeds, and the same program with
the same seeds always results in the same Program Derived Address.

### Seeds

“Seeds” are inputs in the `find_program_address` function, this method provides

> an additional seed called a “bump seed.” The find*program_address method adds a

Reviewer: ```suggestion
an additional seed called a “bump seed.” The find_program_address method adds a
```

Reviewer: For some reason you have this typo everywhere now, please do a find/replace.

> numeric seed called a bump seed that ensures the result is \_off* on the Ed25519

Author: I have just noticed it was an issue with my grammarly configuration changing `""` to `”“` and `'` to `’`  because the markdown file is using both. I will  use `""`  and `'` for consistency. If that's not okay let me know.

curve, ie, is not a valid public key and does not have a corresponding secret
key.

While you, the developer, determine the seeds to pass into the
`find_program_address`, this method provides an additional seed called a “bump
seed.” The cryptographic function for deriving a PDA results in a key that lies
_on_ the Ed25519 curve about 50% of the time. The find*program_address method

> adds a numeric seed called a bump seed that ensures the result \_lies off* on

Author: I had made the changes just that I was waiting for the starter and solutions code to be moved to solana-developers repo so that I can push all changes in one go. Making updates in a few

We can see that the `try_find_program_address` calls the
`create_program_address` method.

```rust
pub fn try_find_program_address(seeds: &[&[u8]], program_id: &Pubkey) -> Option<(Pubkey, u8)> {

    // ...
    for _ in 0..std::u8::MAX {
        {
            // `create_program_address` is called here
            match Self::create_program_address(&seeds_with_bump, program_id) {
               //...
            }
        }
       //...
    }

}

```

### 3. create_program_address

The `create_program_address` method performs a hashing operation over the seeds
and `program_id`. These operations compute a key and verify whether it lies on
the Ed25519 elliptic curve. If a valid PDA is found (i.e., an address that is
_off_ the curve), then either the PDA or an error is returned.

The source code for `create_program_address`:

```rust
pub fn create_program_address(
    seeds: &[&[u8]],
    program_id: &Pubkey,
) -> Result<Pubkey, PubkeyError> {
    if seeds.len() > MAX_SEEDS {
        return Err(PubkeyError::MaxSeedLengthExceeded);
    }
    for seed in seeds.iter() {
        if seed.len() > MAX_SEED_LEN {
            return Err(PubkeyError::MaxSeedLengthExceeded);
        }
    }

>     #[cfg(not(target_os = "solana"))]

Author: I was using the source code from the current solana-program code. I thought that maybe the previous code was using an earlier version of Solana program when the content was written. The code dosen't run on "solana VM " instead it is only run out on other architectures hence the feature flag in current Solana program code. I have removed this and replaced it with comments `//..` to prevent any confusion

`try_find_program_address` method starts with a `bump_seed` of 255, adds it to
the input seeds, and then repeatedly calls the `create_program_address` method
until it finds a valid PDA. Once found, both the PDA and the `bump_seed` are
returned.

Note that different valid bumps generate different valid PDAs for the same input
seeds. The `bump_seed` returned by `find_program_address` will always be the
first valid PDA found.

Using the canonical bump when generating a PDA onchain is crucial. Starting with
a `bump_seed` value of `255` and iterating downward to `0` ensures that the
returned seed will always be the most significant valid `8-bit` value possible.
This `bump_seed` is commonly known as the "_canonical bump_". It's a best
practice always to use the canonical bump and validate every PDA passed into
your program to ensure the integrity of the process.

One point to emphasize is that the `find_program_address` method only returns a
Program-Derived Address and the bump seed used to derive it. The method does not
initialize a new account, nor is any PDA returned by the method necessarily
associated with an account that stores data.

### Use PDA accounts to store data

> Solana programs are stateless, so they manage their state through external
> accounts. Although programs can use the System Program to create non-PDA
> accounts for data storage, PDAs are the choice for storing program-related data.
> This choice is popular because the seeds and canonical bump directly map to the
> same PDA, and the program specified as the program ID can sign on its behalf.

Reviewer: ```suggestion
Solana programs are stateless, so state is stored in separate accounts from where the program's executable is stored. Although programs can use the System Program to create non-PDA
accounts for data storage, PDAs are the choice for storing program-related data.
This choice is popular because the seeds and canonical bump directly map to the
same PDA, and the program specified as the program ID can sign on its behalf.
```

Reviewer: Old language could be clearer.

required mappings into them. It is the natural evolution of the note-taking app
example we just discussed. If you start to use the note creator's public key as
a seed to create one map account per user, then why not use both the creator's
public key and some other known piece of information to derive a PDA for the
note?

We've been mapping seeds to accounts this entire course and have yet to discuss
it explicitly. Think about the Movie Review program we've built in previous
lessons. This program uses a review creator's public key and the title of the
movie they're reviewing to find the address that _should_ be used to store the
review. This approach lets the program create a unique address for every new
review while making it easy to locate a review when needed. When you want to
find a user's review of "Spiderman", you can derive the PDA account's address
using the user's public key and the text "Spiderman" as seeds.

```rust
let (pda, bump_seed) = Pubkey::find_program_address(&[
        initializer.key.as_ref(),
        title.as_bytes().as_ref()
    ],
    program_id
);
```

### Associated token account addresses

> Another practical example of this mapping type is determining the associated
> token account (ATA) addresses. An ATA whose address is derived using a wallet
> address and the mint address of a specific token holds tokens. Solana token
> extensions introduced a new
> [token extensions program ID](https://docs.rs/spl-token-2022/latest/spl_token_2022/fn.id.html).
> You can specify how to create tokens using the legacy token or the new token
> extensions program. Use the new token extensions program, now the standard way
> to create tokens on Solana. The
> [get_associated_token_address_with_program_id](https://docs.rs/spl-associated-token-account/latest/spl_associated_token_account/fn.get_associated_token_address_with_program_id.html)
> function finds the ATA account accepts a `wallet_address`, `token_mint_address`
> and the `token_program_id` as arguments.

Reviewer: ```suggestion
Another practical example of this mapping type is determining associated
token account (ATA) addresses. An ATA is an address used to hold the tokens for a specific account - for example, Jane's USDC account. The ATA address is derived using:

 - the wallet address of the user
 - the mint address of the token
 - the token program used - either the older token program or the newer 
[token extensions program ID](https://docs.rs/spl-token-2022/latest/spl_token_2022/fn.id.html). 

```

I get why it's important to mention token extensions, but let's focus on ATAs for now.

Author: The statement is very short and informative. Thanks

}
```

The mappings between seeds and PDA accounts you use will depend highly on your
specific program. While this isn't a lesson on system design or architecture,
it's worth calling out a few guidelines:

- Use seeds known at the time of PDA derivation
- Be thoughtful about how you group data into a single account
- Be thoughtful about the data structure used within each account
- Simpler is usually better

## Lab

Let's practice with the Movie Review program we've worked on in previous
lessons. No worries if you're jumping into this lesson without doing the last
lesson - it should be possible to follow along either way.

As a refresher, the Movie Review program lets users create movie reviews. These
reviews are stored in an account using a PDA derived from the initializer's
public key and the movie title they are reviewing.

Previously, we finished implementing the ability to update a movie review
securely. In this lab, we'll add the ability for users to comment on a movie
review. We'll use building this feature as an opportunity to work through how to
structure the comment storage using PDA accounts.

### 1. Get the starter code

To begin, you can find

> [the movie program starter code](https://github.com/Unboxed-Software/solana-movie-program/tree/starter)

Reviewer: This repo needs to be updated. https://github.com/Unboxed-Software/solana-movie-program - send PRs to that repo and then I'll marge it and move it to solana-developers

Author: I send a PR a few weeks ago - https://github.com/Unboxed-Software/solana-movie-program/pull/5

Waiting for merge so that I can update the link in the article

Author: I have also updated the solution and main branch today

https://github.com/Unboxed-Software/solana-movie-program/pull/6
https://github.com/Unboxed-Software/solana-movie-program/pull/7

Author: I have updated the frontend solution for this course
https://github.com/Unboxed-Software/solana-movie-frontend/pull/16

Author: Updated the student-intro-program
https://github.com/448-OG/solana-student-intro-program

This has 3 branch changes but all of them upgrading the dependencies and replacing deprecated borsh import. Can the repo just be cloned or should I create 3 PRs for the same thing?

Reviewer: Your student-intro-program repo has been cloned to https://github.com/solana-developers/student-intro-program

Author: > Sorry about the wait BTW, we had in person NYC training then RustConf then Breakpoint.

No problem, guessed you were at Breakpoint. Watched it online you guys are doing awesome work.

    let instruction = MovieInstruction::unpack(instruction_data)?;
    match instruction {
        MovieInstruction::AddMovieReview { title, rating, description } => {

>  			add_movie_review(program_id, accounts, title, rating, description)

Reviewer: Reminder to check your formatting.

        },
        MovieInstruction::UpdateMovieReview { title, rating, description } => {
 			update_movie_review(program_id, accounts, title, rating, description)
        },

        MovieInstruction::AddComment { comment } => {

>  			add_comment(program_id, accounts, comment)

Reviewer: Formatting looks odd.

Reviewer: The formatting wasn't fixed. `update_movie_review(program_id, accounts, title, rating, description)` is indented 16 spaces instead of 4. Please fix formatting before closing!

counter_data.serialize(&mut &mut pda_counter.data.borrow_mut()[..])?;
```

The function initializes two accounts whenever it creates a new review:

1. The first is the review account, which stores the review's contents. This is
   unchanged from the program's version we started with.
2. The second account stores the counter for comments

### 6. Implement add_comment

Finally, implement the `add_comment` function to create new comment accounts.

When creating a new comment for a review, the counter will be incremented on the
comment counter PDA account, and the PDA for the comment account will be derived
using the review address and current count.

Like other instruction processing functions, we'll start by iterating through
accounts passed into the program. Then, before we do anything else, we need to
deserialize the counter account, so we have access to the current comment count:

```rust
pub fn add_comment(
    program_id: &Pubkey,
    accounts: &[AccountInfo],
    comment: String
) -> ProgramResult {

>  msg!("Adding Comment...");

Reviewer: Formatting.
